src/records.jl

```diff
@@ -26,49 +28,91 @@ Simply wraps the value `x` in a `Nullable` and sticks that in an `Attribute` wit
 empty `Function`.
 """
 Attribute(x::T) where {T} = Attribute{T}(() -> x)
+Attribute{T}(x::T) where {T} = Attribute{T}(() -> x)
 
 """
     get(attr::Attribute{T}) -> T
 
 Run set `attr.x` to the output of `attr.f` if `attr.x` is not already set.
 We then return the value stored in `attr.x`
 """
-function Base.get(attr::Attribute)
+function Base.get(attr::Attribute{T}) where T
     if isnull(attr.x)
-        attr.x = Nullable(attr.f())
+        attr.x = Nullable{T}(attr.f())
     end
 
-    return get(attr.x)
+    return get(attr.x)::T
 end
 
+hasfield(T::DataType, name::Symbol) = Base.fieldindex(T, name, false) > 0
+
 """
     Record
 
-Are an `Attribute` container used to store information about a log events including the
-msg, date, level, stacktrace, etc. `Formatter`s use `Records` to format log
+A dictionary-like container with `Symbol` keys used to store information about a log events
+including the msg, date, level, stacktrace, etc. `Formatter`s use `Records` to format log
 message strings.
 
-NOTE: you should access `Attribute`s in a `Record` by using `getindex` (ie: record[:msg])
-as this will correctly extract the value from the `Attribute` container.
+You can access the properties of a `Record` by using `getindex` (ie: record[:msg]).
+
+Subtypes of `Record` should implement `getindex(::MyRecord, ::Symbol)` and key-value pair
+iteration.
+"""
+abstract type Record <: AbstractDict{Symbol, Any} end
+
 """
-abstract type Record end
+    AttributeRecord <: Record
+
+A `Record` which stores its properties as `Attribute`s for lazy evaluation.
+
+Calling `getindex` or iterating will evaluate and cache the properties accessed.
+
+Subtypes of `AttributeRecord` should implement `Memento.getattribute(::MyRecord, ::Symbol)`
+instead of `getindex`.
+"""
+abstract type AttributeRecord <: Record end
+
+level(rec::Record) = rec[:level]
+Base.getindex(rec::Record, attr::Symbol) = getfield(rec, attr)
+Base.haskey(rec::Record, attr::Symbol) = hasfield(rec, attr)
+Base.keys(rec::T) where {T <: Record} = (fieldname(T, i) for i in 1:nfields(rec))
+
+Base.start(rec::Record) = 0
+Base.done(rec::T, state) where {T <: Record} = state >= nfields(T)
+
+function Base.next(rec::T, state::Int) where T <: Record
+    new_state = state + 1
+    return (fieldname(T, new_state) => getfield(rec, new_state), new_state)
+end
+
+function Base.next(rec::T, state::Int) where T <: AttributeRecord
```

Comment:
It might be worthwhile to wrap the iteration stuff in one big
```julia
if isdefined(Base, :iterate) # or a version number
    # iterate
else
    # start, next, done
end
```

Reply:
> Is there a reason we should do this now rather than later?

No, not necessarily.

> Two versions of the same code is precarious IMO.

Yeah that's a fair point. I've just got 0.7 compatibility on the brain. :smile: